Share the block style selector code

scruffian opened the pull request:
Fixes #14466 

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* OpenTable and Calendly both have a block style selector which does essentially the same thing. Let's extract that to a new shared component.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add a Calendly block and check that you can still change the styles in the sidebar
* Add an OpenTable block and check that you can still change the styles in the sidebar
<img width="308" alt="Screenshot 2020-01-17 at 13 48 58" src="https://user-images.githubusercontent.com/275961/72617078-23334f00-3930-11ea-853b-fe7569bf695b.png">
<img width="319" alt="Screenshot 2020-01-17 at 13 48 51" src="https://user-images.githubusercontent.com/275961/72617081-23cbe580-3930-11ea-9ca4-cfe9fcfbf07d.png">


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* no changelog

scruffian committed:
share the block style selector code

scruffian commented on extensions/shared/components/block-styles-selector/style.scss:
```diff
@@ -0,0 +1,21 @@
+		padding-top: 0;
+		position: relative;
+
+		> iframe {
```
this might not be needed anymore

pablinos committed:
Memoized the style previews to prevent them from rerendering

scruffian committed:
remove unused CSS

simison commented:
I'm sure this has been discussed somewhere and I'm not questioning this, but I'm curious why not just use Gutenberg's built-in [`styles`](https://developer.wordpress.org/block-editor/developers/block-api/block-registration/#styles-optional) feature? It also gives similar looking style picker in the sidebar.

scruffian commented:
Gutenberg's built in style switcher only deals with changing the class name for the block, not for changing attributes, which is how this works.

creativecoder commented:
@scruffian 

We're using a similar approach in the Eventbrite block, but using an image rather than a block preview for the "in page" embed option, since that option takes a while to load (during which the preview is blank). https://github.com/Automattic/jetpack/pull/14075/files#diff-67c4a00b3bd3c01b617670a5fd26e417R148

What do you think about making this more generic by using a prop for rendering the preview?

scruffian commented:
We could make it more generic by passing in the preview mechanism as a function. However I'd rather keep this component as close to the core component as possible, with the hope of eventually getting it merged in there. We could look into how to integrate with Eventbrite once this (or that) PR merges...

creativecoder commented:
> However I'd rather keep this component as close to the core component as possible, with the hope of eventually getting it merged in there.

I'd make the same argument in the Core context! The pattern is more broadly usable if it doesn't lock you into a very specific way of rendering a block preview, like it does now. But do what you think is best--what I have now for the Eventbrite block works just fine.

simison commented:
> Gutenberg's built in style switcher only deals with changing the class name for the block, not for changing attributes, which is how this works.

Oh yea, that's annoying limitation indeed. We used style picker in Tiled gallery block but needed the style information in JS as well. There's a helper for that in core and in Jetpack: 

https://github.com/Automattic/jetpack/blob/aa3ad69b50607bda893d7a2b22bf74ae3c81c932/extensions/blocks/tiled-gallery/edit.js#L223

https://github.com/Automattic/jetpack/blob/aa3ad69b50607bda893d7a2b22bf74ae3c81c932/extensions/shared/block-styles.js#L7-L36

By defining styles semantically in the block, you give Gutenberg UI a chance to decide best how to surface that; it might change over time or be surfaced differently in various editor instances (think mobile app version, mobile web version, widgets screen, full site editing mode, writing mode, and generally things we haven't seen yet.)

scruffian committed:
Remove unused class

scruffian commented:
@simison I agree, this needs more thought. I think this PR is still valuable in removing the duplicate code, but we should think more about how we interact with block styles - in a different PR!

simison commented:
> I think this PR is still valuable in removing the duplicate code, but we should think more about how we interact with block styles - in a different PR!

Absolutely! 👍  :-)  Just wanted to highlight other mechanisms in case.

jeherve commented on extensions/shared/components/block-styles-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+ * External Dependencies
+ */
+import classnames from 'classnames';
+import { memo } from 'React';
```
This is breaking things:

```
Module not found: Error: Can't resolve 'React' in '/home/travis/build/Automattic/jetpack/extensions/shared/components/block-styles-selector'
```

Can we do without all of React here? Can we get it from `@wordpress/element` instead?

scruffian replied:
There is a `useMemo` HOC in `@wordpress/element`, but when I use this I get an error from React. Looking into it some more...

scruffian replied:
Ok I added a commit to take care of this too.

pablinos replied:
We might be able to change this to `useMemo` but we need `useSelect` which (as you've seen) also isn't available. It's a shame, but treating it like we have for the changes to `BlockPreview` makes sense to me.

jeherve reviewed (changes requested):
This will need a rebase now, and tests are failing because of the React import.

scruffian committed:
Dont show block previews unless they are supported

scruffian commented:
I rebased this and its working fine. I added a commit to just hide the preview if the version of WordPress doesn't support it:
<img width="297" alt="Screenshot 2020-01-27 at 15 16 09" src="https://user-images.githubusercontent.com/275961/73187000-9733db00-4118-11ea-856f-c90848fba8fe.png">

This seems like the simplest solution for now.

scruffian committed:
use the internal memo if it exists

scruffian commented:
I added a fix for #14466 too.

pablinos commented:
Looks good, and works as expected. It's a shame about the hooks, but never mind.

scruffian committed:
don't try to render the style selector if there's no valid block

jeherve commented on extensions/shared/components/block-styles-selector/index.js:
```diff
@@ -0,0 +1,96 @@
+ * WordPress dependencies
+ */
+import { memo } from '@wordpress/element';
+import { getBlockType, getBlockFromExample, createBlock } from '@wordpress/blocks';
```
`getBlockFromExample` isn't supported on WordPress 5.2, so that won't work:

<img width="307" alt="screenshot 2020-01-29 at 16 38 22" src="https://user-images.githubusercontent.com/426388/73371413-f418c800-42b5-11ea-956b-412b19ad84fb.png">
<img width="323" alt="screenshot 2020-01-29 at 16 38 07" src="https://user-images.githubusercontent.com/426388/73371414-f4b15e80-42b5-11ea-96a2-46d2a2b609b4.png">

Related: #14469

scruffian replied:
@jeherve this is the intended behaviour. For users who don't have these more recent additions to core, we just show a button, as in your screenshot.

jeherve reviewed (approved):
This seems to work well on my end. I'll merge this now, and take care of the matching wpcom reference since it introduces a new file.

Should we now update the Eventbrite block to use this too?

jeherve commented:
Cherry-picked to `branch-8.2` in b878a81080b2ff871dc03f2d3626c5104cf33eac

jeherve commented:
r202313-wpcom

jeherve commented:
> Should we now update the Eventbrite block to use this too?

Logged in #14527
